### ISSUE TITLE
LXD delaying host reboot by 5 minutes

### ISSUE THREAD

**Issue by timlee81** [lxc/lxd#9043]
# Required information

 * Distribution: ubuntu
 * Distribution version: 20.04.2
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.0-80-generic
   * LXC version:
   * LXD version: 4.16 (snap package)
   * Storage backend in use: zfs

# Issue description

Two containers on my system, one with macvlan nictype and the other with routed nictype.
macvlan container is set for autoboot, routed container is started with networkd-dispatcher routable script.

After a boot-up and user login, if I quickly issue a reboot, host doesn't reboot until 5 minutes later. Checking journalctl, I see a 5 minute gap and then the following messages:

Jul 24 12:47:25 odyssey systemd[1]: Stopped xxxx Daemon.
(((gap)))
Jul 24 12:52:20 odyssey ovs-vsctl[9503]: ovs|00001|vsctl|INFO|Called as ovs-vsctl del-port eno1 veth6f14934a
Jul 24 12:52:20 odyssey ovs-vsctl[9503]: ovs|00002|db_ctl_base|ERR|unix:/var/run/openvswitch/db.sock: database connection failed (No such file or directory)

It appears that LXD is internally using OVS for something (it's not installed on the host system but i see binaries in the snap package). The /var/run/openvswitch/db.sock file is not located on my host. If the host has been up for more than 5 minutes, the host will reboot quickly... so I suspect the issue is a boot-up/bring-up issue.

On a fresh-boot, I see the following messages in journalctl (not sure if these provide more info):
Jul 24 13:54:30 odyssey ovs-vsctl[27551]: ovs|00001|db_ctl_base|ERR|missing command name (use --help for help)
...
Jul 25 11:46:19 odyssey ovs-vsctl[1254541]: ovs|00001|vsctl|INFO|Called as ovs-vsctl del-port eno1 vethfe128f2f
Jul 25 11:46:19 odyssey ovs-vsctl[1254541]: ovs|00002|db_ctl_base|ERR|unix:/var/run/openvswitch/db.sock: database connection failed (No such file or directory)

I would have assumed the "del-port" was associated with the tear-down but it appears to be something during boot-up/bring-up.

# Steps to reproduce

 1. Boot host
 2. Login admin user
 3. Reboot host
 4. Wait 5 minutes until reboot occurs

# Information to attach
 -I couldn't find any other relevant logging information.
 -Happy to supply any further helpful information.

I looked around through the source and can see the call-tree... 
network_utils_bridge.go DetachInterface, IsNativeBridge is returning false (I'm guessing lxdbr0 is associated with ovs) which is attempting the del-port (which hangs). Not knowing the lxd architecture and bridge-utilization, I'm surprised it's doing anything with ovs... but I'm guessing it's just a lack of understanding on my part.

**Comment by stgraber** [comment]
Did you try lowering `core.shutdown_timeout`? It defaults to 5min which is how long LXD will wait for existing instances and ongoing operations to complete prior to giving up and killing everything.

**Comment by timlee81** [comment]
"lxc config set core." won't tab-complete core.shutdown_timeout. It looks like it was merged 4.15.
Never the less, the setting did appear to reduce the delay (I set it to one minute, and it took about 1 minute).

It'd probably be better to understand why ovs is being used/invoked at all and avoid any timeout altogether, but I don't have time to dig into the code to determine why.

**Comment by stgraber** [comment]
Can you show `lxc network list`?

**Comment by stgraber** [comment]
It's also worth noting that those errors wouldn't cause a shutdown delay, so there still must be something else keeping LXD from just shutting down and instead hitting the timeout.

**Comment by timlee81** [comment]
>+--------+----------+---------+--------------+--------------------------+-------------+---------+
>|  NAME  |   TYPE   | MANAGED |     IPV4     |           IPV6           | DESCRIPTION | USED BY |
>+--------+----------+---------+--------------+--------------------------+-------------+---------+
>| eno1   | physical | NO      |       (empty)       |   (empty)                       |  (empty)           | 3       |
>+--------+----------+---------+--------------+--------------------------+-------------+---------+
>| lxdbr0 | bridge   | YES     | 10.5.17.1/24 | fd42:cf1:e99b:4fa7::1/64 |             | 1       |
>+--------+----------+---------+--------------+--------------------------+-------------+---------+

Odd- I don't recall using a profile that makes use of the bridge. I do remember copying the default profile...

**Comment by stgraber** [comment]
Can you show `lxc network show lxdbr0`?

**Comment by stgraber** [comment]
@timlee81

**Comment by timlee81** [comment]
Apologies for the delay:

```
config:
  ipv4.address: 10.5.17.1/24
  ipv4.nat: "true"
  ipv6.address: fd42:cf1:e99b:4fa7::1/64
  ipv6.nat: "true"
description: ""
name: lxdbr0
type: bridge
used_by:
- /1.0/profiles/default
managed: true
status: Created
locations:
- none
```

**Comment by stgraber** [comment]
Thanks, that's interesting since there's no mention of openvswitch so it's odd to see it called.
@tomponline any ideas?

**Comment by tomponline** [comment]
If we assume that the call to `ovs-vsctl del-port eno1 vethfe128f2f` is:

1) the cause of the delay (which from what I can tell is not entirely clear yet).
2) and that the call is triggered by LXD (which I'm not 100% sure it is).

Then the suggestion here is that there is a veth interface that is being attempted to be removed from a physical interface rather than an OVS bridge, and that is somehow causing a delay in OVS.

What I would do first is to disable LXD from starting up on boot, restart and quickly reboot again and see if the delay is fixed, at least that way you can correlate LXD to the issue. If its still happening we know its not a LXD issue at least.

So try

```
sudo snap disable lxd
```

And then reboot.

If it still occurs then re-enable LXD with:

```
sudo snap enable lxd
```

And then try stopping all of your instances and then try rebooting and if the issues occurs, that way we can see if its an issue with an LXD network or with a LXD instance.

If the issue then stops then please provide `lxc config show <instance> --expanded` for both your containers.

**Comment by stgraber** [comment]
But as far as we can tell, OVS isn't used at all here.

**Comment by tomponline** [comment]
Yeah indeed, but I'd like to try and find out that LXD is the cause and whether its a LXD network or a LXD instance, as there are some case inside LXD where it would make an ovs call related to bridges when discovering if a bridge is native or OVS type, such as `DetachInterface`. I'm wondering if one of instance `bridged` NICs is incorrectly setup to use `eno1` as its parent, causing `DetachInterface()` to not detect a native bridge and then fallback  to using OVS `BridgePortDelete()`.

**Comment by stgraber** [comment]
Ah yeah, that could explain it indeed.

**Comment by timlee81** [comment]
So it appears the issue is only present when both containers are running.

Debugging:

1. Disabled LXD, reboot: No issue / fast reboot
2. Re-enabled LXD, stopped all instances, reboot: No issue / fast reboot
3. (Both containers active after reboot) Stopped 'routed' nictype container (pihole), reboot: No issue / fast reboot
4. (Both containers active after reboot) Stopped 'macvlan' nictype container (ituneslib), reboot: No issue / fast reboot
5. (Both containers active after reboot) reboot: 5 minute timeout

'ituneslib' config
```
architecture: x86_64
config:
  boot.autostart: "true"
  image.architecture: amd64
  image.description: Debian buster amd64 (20210617_05:24)
  image.os: Debian
  image.release: buster
  image.serial: "20210617_05:24"
  image.type: squashfs
  image.variant: default
  volatile.base_image: 8a93b1554ef3e090e078838e6967448e0ba764a74e7044de1608274d76843109
  volatile.eth0.host_name: macad182ed7
  volatile.eth0.hwaddr: 00:16:3e:f3:32:cb
  volatile.eth0.last_state.created: "false"
  volatile.eth0.name: eth0
  volatile.idmap.base: "0"
  volatile.idmap.current: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.power: RUNNING
  volatile.uuid: c5579084-d213-4d51-a463-fdfe7c0970c9
devices:
  eth0:
    nictype: macvlan
    parent: eno1
    type: nic
  root:
    path: /
    pool: tims_lxd_pool
    type: disk
  timzpoolhost:
    path: /mnt/timzmedia
    source: /timzpool/roshare/Music
    type: disk
ephemeral: false
profiles:
- default
- macvlan_tim
stateful: false
description: ""
```

pihole config:
```
architecture: x86_64
config:
  boot.autostart: "false"
  image.architecture: amd64
  image.description: Debian buster amd64 (20200520_05:24)
  image.os: Debian
  image.release: buster
  image.serial: "20200520_05:24"
  image.type: squashfs
  security.idmap.isolated: "true"
  volatile.base_image: 4ed169fe85fd3fe061252658800382a31c1e1090d6df69ec50bff5d3cfb37235
  volatile.eth0.host_name: veth96670916
  volatile.eth0.hwaddr: 00:16:3e:8c:b6:a2
  volatile.eth0.last_state.created: "false"
  volatile.eth0.name: eth0
  volatile.idmap.base: "1065536"
  volatile.idmap.current: '[{"Isuid":true,"Isgid":false,"Hostid":1065536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":1065536,"Nsid":0,"Maprange":65536}]'
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":1065536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":1065536,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":1065536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":1065536,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.power: RUNNING
  volatile.uuid: 92f9c181-4e6b-49e1-9ca6-5e36332e432d
devices:
  eth0:
    ipv4.address: 192.168.1.9
    ipv6.address: fd4c:af34:7a3d:b6ca::9
    nictype: routed
    parent: eno1
    type: nic
  root:
    path: /
    pool: tims_lxd_pool
    type: disk
ephemeral: false
profiles:
- routed_192.168.1.9
stateful: false
description: ""
```

**Comment by tomponline** [comment]
OK so we are getting closer, but I still don't see any potential link between those two NIC types (macvlan and routed) and OVS.

So we can be 100% certain which NIC is causing the `del-port` command, rather than us rely on a randomly generated `veth....` interface name (as we can see being used in the `del-port` command) please can you set the `host_name` setting on the `routed` NIC using:

```
lxc config device set <instance> eth0 host_name=routedfoo
```

And then retry.

I'm also wondering now if this may be a regression in liblxc (cc @brauner ) as the `routed` NICs are handled by liblxc and maybe there is an OVS call being made inside liblxc itself. There looks like there have been some changes to liblxc's networking recently as I'm also investigating another potential regression over at https://discuss.linuxcontainers.org/t/container-cant-ping-its-neighbour-ip-address-since-its-also-its-own-broadcast-address/11829/7?u=tomp (neither proven yet I would add though).

**Comment by tomponline** [comment]
I've had a look in the liblxc source code and cannot see a reason why it would call del-port either.

**Comment by timlee81** [comment]
Ran: `lxc profile device set routed_192.168.1.9 eth0 host_name=piholelxd` and rebooted several times.

Can no longer reproduce the issue... I thought this would be another debugging step and not a 'fix' :)

**Comment by tomponline** [comment]
@timlee81 how very odd. 

Can you remove the host_name entry to recreate the problem and then enable LXD debug mode so we can gather the system logs to discover where in the process is causing the delay? We need to identify if this is happening in LXD or liblxc.

```
sudo snap set lxd daemon.debug=true; sudo systemctl reload snap.lxd.daemon
```

And then you can see the logs as lxd.daemon in syslog.

**Comment by timlee81** [comment]
@tomponline 

[all.log](https://github.com/lxc/lxd/files/6983621/all.log)
[lxd.log](https://github.com/lxc/lxd/files/6983622/lxd.log)

Unsetting host_name from the profile reproduces the issues (thankfully no heisenbug).

Logs represent full boot-up and shutdown request (with 5 minute hang).

lxd.log is just the lxd.daemon output.
all.log is most of the systems log (filtered out some kernel/audit stuff to aid).

lxd.log contains no calls to ovs/vsctl. I'm assuming if lxd was the culprit, we'd see the call-out to ovs-vsctl in the log.

**Comment by timlee81** [comment]
[success.log](https://github.com/lxc/lxd/files/6984039/success.log)

Managed to get a log for a no-hang scenario. The 'power failure' message is present, but with no 5 minute hang.
Likely some weird timing going on....

**Comment by tomponline** [comment]
Looking at your log files, you have a project called `defaultfovs` - that suggests you have at some point done something with OVS to me :)

```
Aug 13 12:44:11 odyssey lxd.daemon[2001]: t=2021-08-13T12:44:11-0400 lvl=dbug msg="Deleting firewall" driver=bridge network=lxdbr0 project=defaultfovs
```

Can you explain a bit about how that project came about and what you were doing, it may suggest something has been left over.

**Comment by stgraber** [comment]
@timlee81 ping

**Comment by timlee81** [comment]
That doesn't ring a bell (this is a fairly fresh host box).

Regarding the specifics of this box, it's primarily hosting an instance of pi-hole (via routed network type) and a macvlan network type instance for some iTunes Remote library support.

I setup the routed instance via this blog: https://blog.simos.info/how-to-get-lxd-containers-get-ip-from-the-lan-with-routed-network/

I haven't used lxdbr0 explicitly in the host (outside of lxd).

**Comment by stgraber** [comment]
Can you show:
 - lxc project list
 - lxc project show defaultfovs

**Comment by timlee81** [comment]
```
tim@odyssey:~$ lxc project list
+-------------------+--------+----------+-----------------+----------+---------------------+---------+
|       NAME        | IMAGES | PROFILES | STORAGE VOLUMES | NETWORKS |     DESCRIPTION     | USED BY |
+-------------------+--------+----------+-----------------+----------+---------------------+---------+
| default (current) | YES    | YES      | YES             | YES      | Default LXD project | 10      |
+-------------------+--------+----------+-----------------+----------+---------------------+---------+
tim@odyssey:~$ lxc pro
profile  project  
tim@odyssey:~$ lxc project show defaultfovs
Error: Not Found
```

**Comment by stgraber** [comment]
@tomponline could it have been just `default` with the `ovs` from the next line getting mixed in?

**Comment by tomponline** [comment]
> @tomponline could it have been just `default` with the `ovs` from the next line getting mixed in?

Yes possibly.

Looking at the `all.log` again, the fact that the lines:

```
Aug 13 11:18:00 odyssey ovs-vsctl[8381]: ovs|00001|vsctl|INFO|Called as ovs-vsctl del-port eno1 vethd8d8a94e
Aug 13 11:18:00 odyssey ovs-vsctl[8381]: ovs|00002|db_ctl_base|ERR|unix:/var/run/openvswitch/db.sock: database connection failed (No such file or directory)
```

Are occurring before the:

```
Aug 13 11:18:00 odyssey lxd.daemon[1995]: t=2021-08-13T11:18:00-0400 lvl=dbug msg=Handling ip=@ method=GET protocol=unix url="/internal/containers/pihole/onstopns?netns=%2Fproc%2F2272%2Ffd%2F4&project=default&target=stop" username=root
Aug 13 11:18:00 odyssey lxd.daemon[1995]: t=2021-08-13T11:18:00-0400 lvl=dbug msg="Stopping device" device=eth0 instance=pihole instanceType=container project=default type=nic
```

lines, suggests that its liblxc that is triggering the ovs call, before calling into LXD to perform the netns stage clear up.

I need to spend some time in liblxc to address the regression caused by https://github.com/lxc/lxc/commit/365136359f8bf991ed172b498909000ec18b32de so will at the same time try and identify why liblxc is apparently calling ovs when in router mode.

**Comment by tomponline** [comment]
So this PR https://github.com/lxc/lxd/pull/9528 moves way from using liblxc generated veth pairs for `routed` NICs, and instead uses LXD generated veth pairs and liblxc's `phys` NIC type, so this may well fix this issue.

**Comment by stgraber** [comment]
Let's tentatively close this issue as part of that PR then and if @timlee81 sees it again once that's released (4.21), we'll re-open and re-investigate this.
